File: 1415._The_k-th_Lexicographical_String_of_All_Happy_Strings_of_Length_n.cpp
// https: // leetcode.com/problems/construct-smallest-number-from-di-string/description/?envType=daily-question&envId=2025-02-18
// Solution taken from: https://www.youtube.com/watch?v=yyqT5310oJY&ab_channel=Techdose
// Did I understand ? No

class Solution
{
public:
    string getHappyString(int n, int k)
    {
        vector<string> size_n;
        queue<string> q;

        q.push("a");
        q.push("b");
        q.push("c");

        while (!q.empty())
        {
            string curr = q.front();
            q.pop();

            if (curr.size() == n)
            {
                size_n.push_back(curr);
                continue;
            }

            string first = curr;
            string second = curr;

            if (curr.back() == 'a')
            {
                first.push_back('b');
                second.push_back('c');
            }
            else if (curr.back() == 'b')
            {
                first.push_back('a');
                second.push_back('c');
            }
            else
            {
                first.push_back('a');
                second.push_back('b');
            }

            if (size_n.size() >= k)
                break;

            q.push(first);
            q.push(second);
        }
        if (size_n.size() < k)
            return "";
        return size_n[k - 1];
    }
};

File: 1415._The_k-th_Lexicographical_String_of_All_Happy_Strings_of_Length_n.py
# https: // leetcode.com/problems/construct-smallest-number-from-di-string/description/?envType=daily-question&envId=2025-02-18
# Solution taken from: https://www.youtube.com/watch?v=yyqT5310oJY&ab_channel=Techdose
# Did I understand ? No

from collections import deque

class Solution:
    def getHappyString(self, n: int, k: int) -> str:
        size_n = []
        q = deque()

        # Push all start points
        q.append("a")
        q.append("b")
        q.append("c")

        while q:
            curr = q.popleft()
            if len(curr) == n:
                size_n.append(curr)
                continue

            # Build next 2 strings
            last_char = curr[-1]
            if last_char == 'a':
                q.append(curr + 'b')
                q.append(curr + 'c')
            elif last_char == 'b':
                q.append(curr + 'a')
                q.append(curr + 'c')
            else:
                q.append(curr + 'a')
                q.append(curr + 'b')

        if len(size_n) < k:
            return ""
        return size_n[k - 1]